File: services/cms/src/transport/amqp/consumer.go
package amqp

import (
	"callback_service/src/logger"
	"callback_service/src/repository"
	"callback_service/src/service"
	"callback_service/src/transport"
	"context"
	"encoding/json"
	amqp "github.com/rabbitmq/amqp091-go"
)

type IConsumer interface {
	Consume(ctx context.Context) error
}

type Consumer struct {
	amqp    string
	queue   string
	service *service.CallbackService
	logger  logger.ILogger
}

func NewConsumer(amqpURL, queueName string, svc *service.CallbackService, logger logger.ILogger) *Consumer {
	return &Consumer{
		amqp:    amqpURL,
		queue:   queueName,
		service: svc,
		logger:  logger,
	}
}

func (c Consumer) Consume(ctx context.Context) error {
	conn, err := amqp.Dial(c.amqp)
	if err != nil {
		return err
	}
	defer func() {
		if err := conn.Close(); err != nil {
			c.logger.Error("Failed to close connection: %v", err)
		}
	}()

	ch, err := conn.Channel()
	if err != nil {
		return err
	}
	defer func() {
		if err := ch.Close(); err != nil {
			c.logger.Error("Failed to close channel: %v", err)
		}
	}()

	q, err := transport.DeclareQueue(ch, c.queue)
	if err != nil {
		return err
	}

	msgs, err := ch.Consume(
		q.Name, // queue
		"",     // consumer
		true,   // auto-ack
		false,  // exclusive
		false,  // no-local
		false,  // no-wait
		nil,    // args
	)
	if err != nil {
		return err
	}

	forever := make(chan bool)

	go func() {
		for d := range msgs {
			c.logger.Info("Received a message: %s", d.Body)

			var msg repository.Callback
			if err := json.Unmarshal(d.Body, &msg); err != nil {
				c.logger.Error("failed to unmarshal message: %v", err)
				continue
			}

			if err := c.service.CreateCallback(ctx, msg.Name, msg.Phone, msg.Type, msg.Idea); err != nil {
				c.logger.Error("failed to create callback: %v", err)
			}

			producer := NewProducer(c.amqp, c.queue, c.logger)
			if err := producer.Produce(ctx, "new callback"); err != nil {
				c.logger.Error("failed to send notification: %v", err)
			}
		}
	}()

	c.logger.Info("Waiting for messages. To exit press CTRL+C")
	<-forever

	return nil
}

File: services/tg_bot/app/core/transport/amqp/consumer.py
import aio_pika
import asyncio
from app.config.config import config
import json

class Consumer:
    def __init__(self, bot, amqp_url: str, queue_name: str):
        self.bot = bot
        self.amqp_url = amqp_url
        self.queue_name = queue_name

    async def on_message(self, message: aio_pika.IncomingMessage):
        async with message.process():
            print(f" [x] Received message {message.body.decode()}")
            await self.bot.send_message(config['bot']['chat_id'], "Новый callback")

    async def consume(self):
        connection = await aio_pika.connect_robust(self.amqp_url)
        async with connection:
            channel = await connection.channel()
            queue = await channel.declare_queue(self.queue_name, durable=True)
            await queue.consume(self.on_message)
            print(" [*] Waiting for messages. To exit press CTRL+C")
            await asyncio.Future()  # Run forever
